### PR TITLE
feat: Disable Introspection in Production

### PR DESCRIPTION
**Pull request by fahimalizain**
Introspection is Disabled based on the following conditions in site_config.json
```json
{
  "developer_mode": 0,
  "enable_introspection_in_production": 0
}
```

### CODE REVIEW

**Review by assemmarwan — APPROVED** [review]

File: frappe_graphql/utils/middlewares/disable_introspection_queries.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+def disable_introspection_queries(next_resolver, obj, info: GraphQLResolveInfo, **kwargs):
+    # https://github.com/jstacoder/graphene-disable-introspection-middleware
+    if is_introspection_disabled() and info.field_name.lower() in ['__schema', '__introspection']:
+        raise IntrospectionDisabled("Introspection is disabled")
+
+    return next_resolver(obj, info, **kwargs)
+
+
+def is_introspection_disabled():
+    return not frappe.local.conf.get("developer_mode") and \
```
Comment (e-lobo) [inline frappe_graphql/utils/middlewares/disable_introspection_queries.py]: cint(frappe.local.conf.get("developer_mode")) & cint(frappe.local.conf.get("enable_introspection_in_production"))?

File: frappe_graphql/utils/middlewares/disable_introspection_queries.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+
+class IntrospectionDisabled(Exception):
+    pass
+
+
+def disable_introspection_queries(next_resolver, obj, info: GraphQLResolveInfo, **kwargs):
+    # https://github.com/jstacoder/graphene-disable-introspection-middleware
+    if is_introspection_disabled() and info.field_name.lower() in ['__schema', '__introspection']:
+        raise IntrospectionDisabled("Introspection is disabled")
```
Comment (e-lobo) [inline frappe_graphql/utils/middlewares/disable_introspection_queries.py]:
```suggestion
        raise IntrospectionDisabled(_("Introspection is disabled"))
```

File: frappe_graphql/graphql.py [hunk 3]
```diff
@@ -13,6 +14,7 @@ def execute(query=None, variables=None, operation_name=None):
         variable_values=variables,
         operation_name=operation_name,
         field_resolver=default_field_resolver,
+        middleware=[disable_introspection_queries],
```
Comment (e-lobo) [inline frappe_graphql/graphql.py:17]: maybe we can allow users to extend middle-wares through hooks like we do others ?

Comment (e-lobo) [reply frappe_graphql/graphql.py:17]: @assemmarwan ?

**Comment by e-lobo** [comment]
@fahimalizain please do add to docs as well.

File: frappe_graphql/graphql.py (hunk 3 above)
Comment (assemmarwan) [reply frappe_graphql/graphql.py:17]: Good idea 👍🏻 , yet I think we can have that part of another PR

File: frappe_graphql/utils/middlewares/disable_introspection_queries.py (hunk 1 above)
Comment (fahimalizain) [reply frappe_graphql/utils/middlewares/disable_introspection_queries.py]: @e-lobo fixed 👍🏼

File: frappe_graphql/utils/middlewares/disable_introspection_queries.py (hunk 2 above)
Comment (fahimalizain) [reply frappe_graphql/utils/middlewares/disable_introspection_queries.py]: @e-lobo fixed 👍🏼

**Review by e-lobo — APPROVED** [review]

**Pull request merged**